pkg/openapi/validation.go

```diff
@@ -0,0 +1,240 @@
+package openapi
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/nirmata/kyverno/data"
+
+	"github.com/golang/glog"
+
+	"github.com/nirmata/kyverno/pkg/engine"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	v1 "github.com/nirmata/kyverno/pkg/api/kyverno/v1"
+
+	openapi_v2 "github.com/googleapis/gnostic/OpenAPIv2"
+	"github.com/googleapis/gnostic/compiler"
+	"k8s.io/kube-openapi/pkg/util/proto"
+	"k8s.io/kube-openapi/pkg/util/proto/validation"
+
+	"gopkg.in/yaml.v2"
+)
+
+var openApiGlobalState struct {
+	mutex                sync.RWMutex
+	document             *openapi_v2.Document
+	definitions          map[string]*openapi_v2.Schema
+	kindToDefinitionName map[string]string
+	crdList              []string
+	models               proto.Models
+}
+
+func init() {
+	defaultDoc, err := getSchemaDocument()
+	if err != nil {
+		panic(err)
+	}
+
+	err = useOpenApiDocument(defaultDoc)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func ValidatePolicyMutation(policy v1.ClusterPolicy) error {
+	openApiGlobalState.mutex.RLock()
+	defer openApiGlobalState.mutex.RUnlock()
+
+	var kindToRules = make(map[string][]v1.Rule)
+	for _, rule := range policy.Spec.Rules {
+		if rule.HasMutate() {
+			rule.MatchResources = v1.MatchResources{
+				UserInfo: v1.UserInfo{},
+				ResourceDescription: v1.ResourceDescription{
+					Kinds: rule.MatchResources.Kinds,
+				},
+			}
+			rule.ExcludeResources = v1.ExcludeResources{}
+			for _, kind := range rule.MatchResources.Kinds {
+				kindToRules[kind] = append(kindToRules[kind], rule)
+			}
+		}
+	}
+
+	for kind, rules := range kindToRules {
+		newPolicy := *policy.DeepCopy()
+		newPolicy.Spec.Rules = rules
+		resource, _ := generateEmptyResource(openApiGlobalState.definitions[openApiGlobalState.kindToDefinitionName[kind]]).(map[string]interface{})
```

Comment:
Does this `generateEmptyResource` fill in the default value for each field? As previously we encountered the issue described [here](https://github.com/nirmata/kyverno/issues/522#issuecomment-590640851), when the policy is defined with variable context, it throws error as the referred path is not present.

Reply:
Yes, if the openApi schema does have a default value for the field - it will use that over an empty value.

I have added the forcemutate function which does not check for any conditions and simply mutates the given resource.

Any substitution variables {{suchAsThese}}, will be replaced with dummy values - and does not check any conditions before mutating. Hence if there is a rule which wants to mutate resource if it has a specific name(which would fail since empty resource has only empty values) - forcemutate ensures that it will do the mutation without checking the condition.

The patched resource is then validated.